r2i/tflite/engine.cc

```diff
@@ -230,5 +226,100 @@ Engine::~Engine () {
   this->Stop();
 }
 
+template <typename T>
+static uint8_t ConvertToFixedPoint(const T value, const TfLiteTensor &tensor) {
+  return (value / tensor.params.scale) + tensor.params.zero_point;
+}
+
+template <typename T>
+static T ConvertToFloatingPoint(const uint8_t value,
+                                const TfLiteTensor &tensor) {
+  return (value - tensor.params.zero_point) * tensor.params.scale;
+}
+
+void Engine::SetupResolver(::tflite::ops::builtin::BuiltinOpResolver
+                           &/*resolver*/) {
+  // No implementation for tflite engine
+}
+
+void Engine::SetInterpreterContext(::tflite::Interpreter */*interpreter*/) {
+  // No implementation for tflite engine
+}
+
+void Engine::PreprocessInputData(const float *input_data, const int size,
+                                 ::tflite::Interpreter *interpreter, r2i::RuntimeError &error) {
+  const auto &input_indices = interpreter->inputs();
+  const auto *tensor = interpreter->tensor(input_indices[0]);
+
+  if (!input_data) {
+    error.Set (RuntimeError::Code::WRONG_API_USAGE, "Failed to get image data");
+    return;
+  }
+
+  if (kTfLiteUInt8 == tensor->type) {
+    auto input_fixed_tensor = interpreter->typed_input_tensor<uint8_t>(0);
+
+    // Convert to fixed point and write the data to the input tensor
+    for (int index = 0; index < size; index++) {
+      input_fixed_tensor[index] = ConvertToFixedPoint(input_data[index], *tensor);
+    }
+  } else if (kTfLiteFloat32 == tensor->type) {
+    auto input_tensor = interpreter->typed_tensor<float>(input_indices[0]);
+
+    memcpy(input_tensor, input_data, size * sizeof(float));
```

Comment:
Do we really need to copy the data? can t we pass it in directly?

Reply:
We could do a for loop to write directly to the input tensor, but I would say that it would be slower.